.github/workflows/test-build-deploy.yml

```diff
@@ -338,17 +338,42 @@ jobs:
         uses: actions/download-artifact@v4
         with:
           name: Race-enabled Mimir
-      - name: Load race-enabled mimir into Docker
+      - name: Check Whether In Deploy Process
+        id: check_deploy
         run: |
-          docker load -i ./mimir_race_image
-          export IMAGE_TAG_RACE=$(make image-tag-race)
-          docker run "grafana/mimir:$IMAGE_TAG_RACE" --version
+          if [[ "${{ startsWith(github.ref, 'refs/tags/') || startsWith(github.ref, 'refs/heads/r') }}" == 'true' && "${{ github.event_name == 'push' }}" == 'true' && "${{ github.repository == 'grafana/mimir' }}" == 'true' ]]; then
+            echo "isDeploy=true" >> $GITHUB_OUTPUT
+          else
+            echo "isDeploy=false" >> $GITHUB_OUTPUT
+          fi
       - name: Preload Images
         # We download docker images used by integration tests so that all images are available
         # locally and the download time doesn't account in the test execution time, which is subject
         # to a timeout
         run: go run ./tools/pre-pull-images | xargs -n1 -P4 docker pull
+      - name: Load race-enabled mimir-distroless into Docker
+        run: |
+          docker load -i ./mimir_race_image
+          export IMAGE_TAG_RACE=$(make image-tag-race)
+          docker run "grafana/mimir-distroless:$IMAGE_TAG_RACE" --version
+      - name: Integration Tests With Distroless Docker Image
+        run: |
+          export IMAGE_TAG_RACE=$(make image-tag-race)
+          export MIMIR_IMAGE="grafana/mimir-distroless:$IMAGE_TAG_RACE"
+          export IMAGE_TAG=$(make image-tag)
+          export MIMIRTOOL_IMAGE="grafana/mimirtool:$IMAGE_TAG"
+          export MIMIR_CHECKOUT_DIR="/go/src/github.com/grafana/mimir"
+          echo "Running integration tests with image: $MIMIR_IMAGE (Mimir), $MIMIRTOOL_IMAGE (Mimirtool)"
+          echo "Running integration tests (group ${{ matrix.test_group_id }} of ${{ matrix.test_group_total }}) with Go version: $(go version)"
+          ./.github/workflows/scripts/run-integration-tests-group.sh --index ${{ matrix.test_group_id }} --total ${{ matrix.test_group_total }}
+      - name: Load race-enabled mimir into Docker
+        if: ${{ steps.check_deploy.outputs.isDeploy == 'true' }}
+        run: |
+          docker load -i ./mimir_race_image
+          export IMAGE_TAG_RACE=$(make image-tag-race)
+          docker run "grafana/mimir:$IMAGE_TAG_RACE" --version
       - name: Integration Tests
+        if: ${{ steps.check_deploy.outputs.isDeploy == 'true' }}
```

Comment:
Why not make the image a gh workflow variable and pass it into this step so that we don't have to repeat the integration test section?

Reply:
what do you mean by gh workflow variable? depends on whether it is deployment or not set images=[alpine, distroless] or just images=[distroless]? my problem with variable is to decode a list is a bit tricky. or you mean something else?

Reply:
Yes, that is what I mean. Either that or a [matrix](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs).

I think because everywhere it is used is a variable it should be clear what is changing. It deduplicates the job that we have to maintain.

Reply:
I am thinking right now docker run mimir and integration tests are in 2 different steps, if we want to deduplicates the job, we need merge the 2 steps together. If this is not blocking how about we merge it as it is and I can do another refectory of existing steps to minimise the changes.

Reply:
rewrite to a bash scripts to avoid duplication, in this way we keep the sequential execution order since the tests use mimir_image environement variable.